pkg/controller/statusmanager/pod_status.go

```diff
@@ -0,0 +1,278 @@
+		} else if ds.Status.NumberAvailable == 0 { // NOTE: update this if we ever expect empty (unscheduled) daemonsets ~cdc
+			progressing = append(progressing, fmt.Sprintf("DaemonSet %q is not yet scheduled on any nodes", dsName.String()))
+			dsProgressing = true
+		} else if ds.Generation > ds.Status.ObservedGeneration {
+			progressing = append(progressing, fmt.Sprintf("DaemonSet %q update is being processed (generation %d, observed generation %d)", dsName.String(), ds.Generation, ds.Status.ObservedGeneration))
+			dsProgressing = true
+		}
+
+		if dsProgressing || ds.Annotations["release.openshift.io/version"] != targetLevel {
+			reachedAvailableLevel = false
+		}
+
+		dsState, ok := lastState.daemonsetState[dsName]
+		if ok && !reflect.DeepEqual(dsState.LastSeenStatus, ds.Status) {
+			ok = false // invalidate
+		}
+		if !ok {
+			dsState.Name = dsName
+			dsState.LastChangeTime = time.Now()
+			ds.Status.DeepCopyInto(&dsState.LastSeenStatus)
+			lastState.daemonsetState[dsName] = dsState
+		}
+
+		// Catch hung rollouts
+		if dsProgressing && ok && (time.Since(dsState.LastChangeTime)) > ProgressTimeout {
+			hung = append(hung, fmt.Sprintf("DaemonSet %q rollout is not making progress - last change %s", dsName.String(), dsState.LastChangeTime.Format(time.RFC3339)))
+		}
+	}
+
+	for _, depName := range status.deployments {
+		dep := &appsv1.Deployment{}
+		if err := status.client.Get(context.TODO(), depName, dep); err != nil {
+			log.Printf("Error getting Deployment %q: %v", depName.String(), err)
+			progressing = append(progressing, fmt.Sprintf("Waiting for Deployment %q to be created", depName.String()))
+			// Assume the OperConfig Controller is in the process of reconciling
+			// things; it will set a Degraded status if it fails.
+			continue
+		}
+
+		depProgressing := false
+
+		if dep.Status.UnavailableReplicas > 0 {
+			progressing = append(progressing, fmt.Sprintf("Deployment %q is not available (awaiting %d nodes)", depName.String(), dep.Status.UnavailableReplicas))
+			depProgressing = true
+		} else if dep.Status.AvailableReplicas == 0 {
+			progressing = append(progressing, fmt.Sprintf("Deployment %q is not yet scheduled on any nodes", depName.String()))
+			depProgressing = true
+		} else if dep.Status.ObservedGeneration < dep.Generation {
+			progressing = append(progressing, fmt.Sprintf("Deployment %q update is being processed (generation %d, observed generation %d)", depName.String(), dep.Generation, dep.Status.ObservedGeneration))
+			depProgressing = true
+		}
+
+		depState, ok := lastState.deploymentState[depName]
+		if ok && !reflect.DeepEqual(depState.LastSeenStatus, dep.Status) {
+			ok = false // invalidate
+		}
+		if !ok {
+			depState.Name = depName
+			depState.LastChangeTime = time.Now()
+			dep.Status.DeepCopyInto(&depState.LastSeenStatus)
+			lastState.deploymentState[depName] = depState
+		}
+
+		// Catch hung rollouts
+		if depProgressing && ok && (time.Since(depState.LastChangeTime)) > ProgressTimeout {
```

Comment:
as above you could move the `if depProgressing` earlier, and you need to do cleanup